### ISSUE TITLE
IPython.display.Audio not working for WAV audio

### ISSUE THREAD

**Issue by mcmar** [microsoft/vscode-jupyter#1012]
# Bug: Notebook Editor, Interactive Window, Editor cells

Issue has previously been discussed, but I have new information that is relevant. I'm forced to create a new issue because all previous issues have been closed and locked.
Please do not close it without considering this new information and responding to it.
I'm only looking for support for uncompressed wav audio format. No need for mp3, mpeg, ogg, or anything else.
New information:
1. Previous discussion with @rchiodo was limited to supporting ffmpeg codec. Support for wav audio files is not dependent in any way on ffmpeg codec. Wav audio can be played back without any codec. No dependency on ffmpeg exists for playing wav audio, which is the most common type used for machine learning because it can be sampled and reconstructed easily (ie wavenet/wavernn).
2. Microsofter and VSCode creater, @bpasero, claims that sound may be played back without DOM support via node.js https://github.com/Microsoft/vscode/issues/421 This would eliminate previous reasons for closing this issue in the past.
3. @mjbvz says that ffmpeg codec is unsupported by the core VSCode, but:
a) does not state that ffmpeg cant be supported by an extension like vscode-python
b) does not state that any codec is required in order to play wav audio files
https://github.com/microsoft/vscode/issues/66050
4. @mjbvz again states here that only certain types of media are unsupported and mentions a dependency on ffmpeg https://github.com/microsoft/vscode/issues/32540
He does not state that all audio playback is unsupported.
## Steps to cause the bug to occur
Type the following code in VSCode, and run the cell (e.g. using SHIFT+ENTER):
```
import numpy as np
import IPython
T = 2.0    # seconds
sr = 22050 # sample rate
t = np.linspace(0, T, int(T*sr), endpoint=False) # time variable
x = 0.5*np.sin(2*np.pi*440*t)                # pure sine wave at 440 Hz
IPython.display.Audio(x, rate=sr)
```
You will get a disabled player. The same code in the browser will render a usable player.
## Actual behavior
IPython.display.Audio renders a disabled player:
![image](https://user-images.githubusercontent.com/8251430/83338915-b042c900-a27d-11ea-9057-e3fdda56af52.png)

## Expected behavior
A valid, playable audio control should be rendered, same as in the browser:
![image](https://user-images.githubusercontent.com/8251430/83338948-ebdd9300-a27d-11ea-9be2-2ba49acb5e4f.png)

### Your Jupyter and/or Python environment

_Please provide as much info as you readily know_

-   **Jupyter server running:** Local
-   **Extension version:** 2020.5.80290
-   **VS Code version:** 1.45.1
-   **Setting python.jediEnabled:** true
-   **Setting python.languageServer:** Microsoft
-   **Python and/or Anaconda version:** 3.7.6
-   **OS:** Linux (Ubuntu 18.04 LTS) 
-   **Virtual environment:** conda

## Developer Tools Console Output
Far too big to fit here. No relevant information. Just deprecation warnings and info output.

Microsoft Data Science for VS Code Engineering Team: @rchiodo, @IanMatthewHuff, @DavidKutu, @DonJayamanne, @greazer, @joyceerhl

**Comment by portogasDace** [comment]
I too have this bug can't have a work around?

**Comment by rchiodo** [comment]
Sorry the work around would be to write your own transform for text/html to create your own audio control that uses the WAV data without FFMpeg

**Comment by kirianguiller** [comment]
Hi everyone, the WAV reader still seems to be not working for vscode jupyter notebook. Is there any plan to fix it in a near future ?

**Comment by rchiodo** [comment]
It's unlikely thi will be fixed. It's by design:
https://github.com/microsoft/vscode/issues/32540#issuecomment-381847598

We'd have to write something to hack it to get it to work and I don't see that happening unless we get a LOT of upvotes on getting audio for WAV to work.

**Comment by clouds56** [comment]
Maybe we should first leave this open as it doesn't get fixed.

**Comment by matangover** [comment]
Hi all, I've found a workaround using the Web Audio API. Save the following into a file called `vscode_audio.py`:

```python
import IPython.display
import numpy as np
import json

def Audio(audio: np.ndarray, sr: int):
    """
    Use instead of IPython.display.Audio as a workaround for VS Code.
    `audio` is an array with shape (channels, samples) or just (samples,) for mono.
    """

    if np.ndim(audio) == 1:
        channels = [audio.tolist()]
    else:
        channels = audio.tolist()

    return IPython.display.HTML("""
        <script>
            if (!window.audioContext) {
                window.audioContext = new AudioContext();
                window.playAudio = function(audioChannels, sr) {
                    const buffer = audioContext.createBuffer(audioChannels.length, audioChannels[0].length, sr);
                    for (let [channel, data] of audioChannels.entries()) {
                        buffer.copyToChannel(Float32Array.from(data), channel);
                    }
            
                    const source = audioContext.createBufferSource();
                    source.buffer = buffer;
                    source.connect(audioContext.destination);
                    source.start();
                }
            }
        </script>
        <button onclick="playAudio(%s, %s)">Play</button>
    """ % (json.dumps(channels), sr))
```

In your notebook, use:
```python
from vscode_audio import Audio
Audio(audio, sr)
```
where `audio` in a numpy array with shape (channels, samples) or just (samples,) for mono.

Obviously this is quite hacky but works for me...
I suggest to reopen this issue.

**Comment by robinhad** [comment]
Hello all, I started working on workaround - extracting audio tracks (wav only) from HTML DOM and playing them using Javascript.
I'll keep updates here, planning to make a pull request.
<img width="661" alt="зображення" src="https://user-images.githubusercontent.com/5759207/109562477-1dbea700-7ae7-11eb-8fac-f7a08be4d5c0.png">

**Comment by robinhad** [comment]
Implemented independent playback for different files and start/stop on-demand.
https://user-images.githubusercontent.com/5759207/109793490-624c5e80-7c1d-11eb-9bca-31770ca971b5.mov

**Comment by robinhad** [comment]
Hi all, I made a pull request that fixes this issue https://github.com/microsoft/vscode-notebook-renderers/pull/33.
Could you open this issue for the time while this pull request is reviewed?

**Comment by DonJayamanne** [comment]
We have discuss this issue with VS Code, and will wait to see how or what VS Code does to address this.

**Comment by hgf777-br** [comment]
Hi, I will apreciate too to have this issue resolved. It will be nice to play the sounds directly in Vscode instead of going out to the system every time.

**Comment by leoauri** [comment]
I, too, would really appreciate this being implemented.  This is blocking me from using the jupyter extension

**Comment by rchiodo** [comment]
Here's the active issue on VS code:
https://github.com/microsoft/vscode/issues/118275

Please upvote that issue.

**Comment by leoauri** [comment]
Thank you @rchiodo

**Comment by mbadawi23** [comment]
I too need this... This is my +1 vote.

**Comment by jkquijas** [comment]
I also would like to see this

**Comment by jingw222** [comment]
Still not fixed yet.

**Comment by BazB179** [comment]
Bruh it's still not fixed... come on

**Comment by dishkakrauch** [comment]
Any news?)

**Comment by rchiodo** [comment]
This isn't something the jupyter extension is going to address. It's a VS code issue. You might track this instead:

https://github.com/microsoft/vscode/issues/118275

**Comment by BazB179** [comment]
So both parties will blame each other and nothing will be done? What happened to building with ffmpeg?

**Comment by rchiodo** [comment]
> So both parties will blame each other and nothing will be done? What happened to building with ffmpeg?

You should probably ask on the VS code issue. We're not going to include ffmpeg in the jupyter extension but it does seem like WAV files should be playable. I'm guessing it's a limitation of electron and how it bundles support for audio/video.

**Comment by BazB179** [comment]
Maybe they will hear if you give feedback in this: https://www.research.net/r/DB6HGK9

**Comment by p-i-** [comment]
^ I've created a feature-request for a barebones functionality to play a mono/stereo np.array from a cell output.
I think this should be addressed separately from the much larger task of a fully-fledged media player.
If the team would care to provide this, it would give us audio engineers enough to work with.
Forever grateful for the tremendous work of the VSCode team, and in particular the .ipynb unit. It's a splendid technology.

(I've posted the same message in https://github.com/microsoft/vscode/issues/118275 which appears to be the nexus for this subject. It seems to me that the smart / practical first move is to first provide a minimal implementation that plays a raw audio-buffer from cell-output. Whatever obstacles are blocking a full implementation of the browser-based Jupyter Notebook cell-output-play-audio functionality, we can assume they are significant, as no plan to fix has been announced despite massive upvoting. So I propose this as a stop-gap.

**Comment by Selimonder** [comment]
The issue persists with 1.67.2 version.

**Comment by IavTavares** [comment]
Thanks @lukepighetti  I've just upvoted the issue. ;)

**Comment by InfernalWAVE** [comment]
looks like it is at 26 votes and it needed 20 to get on the backlog, now we wait.

**Comment by P-Light** [comment]
> Ticket https://github.com/microsoft/vscode/issues/151149 has been selected as candidate for backlog

[VSCode moved it to their backlog as of July 2nd. Progress!](https://github.com/microsoft/vscode/issues/118275#issuecomment-1173073070) 🎉

**Comment by austinmw** [comment]
Hi, what's the latest ticket for `IPython.display.Video` support as well so I can upvote?

**Comment by amunger** [comment]
@austinmw - video also depends the same core issue - https://github.com/microsoft/vscode/issues/118275
As [the latest comment](https://github.com/microsoft/vscode/issues/118275#issuecomment-1213064316) states, please comment over there if your specific scenario doesn't work after the codecs are available.

**Comment by berkeleymalagon** [comment]
Excited for this feature. Happy to help however I can.

**Comment by SaminYaser-work** [comment]
Looks like its fixed now in 1.71 (August 2022)


![image](https://user-images.githubusercontent.com/40205522/188491177-3c2552e7-8dbd-4bdf-ac95-7ba9c6ae6c0e.png)

**Comment by amunger** [comment]
thanks for the confirmation! Please feel free to file any bugs you find as new issues
